### PR TITLE
K-core implementation for undirected graphs

### LINKED ISSUE #1909: [FEA] MNMG K-core
**Is your feature request related to a problem? Please describe.**
It is common to first remove degree 0 & 1 vertices (in a recursive manner as removing a degree 1 vertex can create another degree 1 vertex) to efficiently enumerate triangles. MNMG K-core is a pre-requisite to implement efficient MNMG triangle counting.

**Describe the solution you'd like**
I think K-core can be implemented using the ```update_frontier_v_push_if_out_nbr``` graph primitive (https://github.com/rapidsai/cugraph/blob/branch-21.12/cpp/include/cugraph/prims/update_frontier_v_push_if_out_nbr.cuh#L801).

Starting with the vertices with their degrees [1, K) as the initial frontier, decrement degrees for the vertices in the next level, and push the vertices in the queue for the iteration in the next step if vertex degrees become smaller than K.

### PR DESCRIPTION
**Pull request by seunghwak**
Close #1909.

This PR aims to implement K-core only for undirected graphs.

- [x] Implement K-core implementation
- [x] Test/debug on SG
- [x] Test/debug on MG

### CODE REVIEW

**Review by ChuckHastings — APPROVED** [review]
One observation, willing to merge without the change.

File: cpp/src/cores/core_number_impl.cuh [hunk 1]
```diff
@@ -25,16 +30,255 @@
...
+                    (degree_type == k_core_degree_type_t::OUT) ||
+                    (degree_type == k_core_degree_type_t::INOUT),
+                  "Invalid input argument: degree_type should be IN, OUT, or INOUT.");
+  CUGRAPH_EXPECTS(k_first <= k_last, "Invalid input argument: k_first <= k_last.");
+
+  if (do_expensive_check) {
+    if (graph_view.is_multigraph()) {
+      CUGRAPH_FAIL("unimplemented.");  // check for multi-edges
+    }
+    CUGRAPH_FAIL("unimplemented.");  // check for self-loops
```
[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/src/cores/core_number_impl.cuh:75]: Yes, I should.

[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/src/cores/core_number_impl.cuh]: Done.

**Review by kaatish — APPROVED** [review]

**Comment by ChuckHastings** [comment]
@gpucibot merge

**Pull request merged**